File: src/chain-observers/filters/impacted-account-filter.ts
import type { WorkerBee } from "../../bot";
import { ImpactedAccountClassifier } from "../classifiers";
import type { TRegisterEvaluationContext } from "../classifiers/collector-classifier-base";
import type { TFilterEvaluationContext } from "../factories/data-evaluation-context";
import { FilterBase } from "./filter-base";

export class ImpactedAccountFilter extends FilterBase {
  public constructor(
    worker: WorkerBee,
    accounts: string[]
  ) {
    super(worker);

    this.accounts = new Set(accounts);
  }

  private readonly accounts: Set<string>;

  public usedContexts(): Array<TRegisterEvaluationContext> {
    return [
      ImpactedAccountClassifier
    ];
  }

  public async match(data: TFilterEvaluationContext): Promise<boolean> {
    const impactedAccount = await data.get(ImpactedAccountClassifier);

    for(const account of this.accounts)
      if(impactedAccount.impactedAccounts[account] !== undefined)
        return true;

    return false;
  }
}

File: python/workerbee/chain_observers/filters/impacted_account_filter.py
"""ImpactedAccountFilter — matches when specified accounts are impacted.

Mirrors src/chain-observers/filters/impacted-account-filter.ts.

Uses ImpactedAccountClassifier.
"""

from __future__ import annotations

from typing import TYPE_CHECKING

from .._ordered_set import OrderedSet
from .filter_base import FilterBase

if TYPE_CHECKING:
    from ..classifiers.collector_classifier_base import TRegisterEvaluationContext
    from ..factories.data_evaluation_context import DataEvaluationContext


class ImpactedAccountFilter(FilterBase):
    def __init__(self, accounts: list[str]) -> None:
        super().__init__()
        self._accounts: OrderedSet[str] = OrderedSet(accounts)

    def used_contexts(self) -> list[TRegisterEvaluationContext]:
        from ..classifiers.impacted_account_classifier import ImpactedAccountClassifier

        return [ImpactedAccountClassifier]

    async def match(self, data: DataEvaluationContext) -> bool:
        from ..classifiers.impacted_account_classifier import ImpactedAccountClassifier

        result = await data.get(ImpactedAccountClassifier)

        return any(result["impacted_accounts"].get(account) is not None for account in self._accounts)
